Source/BoundaryConditions/PML.cpp

```diff
@@ -515,6 +517,42 @@ PML::PML (const BoxArray& grid_ba, const DistributionMapping& /*grid_dm*/,
         WarpX::GetInstance().getHfield_fp(0,2).ixType().toIntVect() ), dm, 2, ngb );
 #endif
 
+    if (WarpX::em_solver_medium == MediumForEM::Macroscopic) {
+        // Allocating macroproperties in pml at cell-centers
+        pml_eps_fp = std::make_unique<MultiFab>(ba, dm, 1, nge);
+        pml_mu_fp = std::make_unique<MultiFab>(ba, dm, 1, nge);
+        pml_sigma_fp = std::make_unique<MultiFab>(ba, dm, 1, nge);
+
+        // Initializing macroparameter multifab //
+        auto& warpx = WarpX::GetInstance();
+        auto& macroscopic_properties = warpx.m_macroscopic_properties;
+
+        // Initialize sigma, conductivity
+        if (macroscopic_properties->m_sigma_s == "constant") {
+            pml_sigma_fp->setVal(macroscopic_properties->m_sigma);
+        } else if (macroscopic_properties->m_sigma_s == "parse_sigma_function") {
+            macroscopic_properties->InitializeMacroMultiFabUsingParser(pml_sigma_fp.get(),
+                getParser(macroscopic_properties->m_sigma_parser), lev);
+        }
+
+        // Initialize epsilon, permittivity
+        if (macroscopic_properties->m_epsilon_s == "constant") {
+            pml_eps_fp->setVal(macroscopic_properties->m_epsilon);
+        } else if (macroscopic_properties->m_epsilon_s == "parse_epsilon_function") {
+            macroscopic_properties->InitializeMacroMultiFabUsingParser(pml_eps_fp.get(),
+                getParser(macroscopic_properties->m_epsilon_parser), lev);
+        }
+
+        // Initialize mu, permeability
+        if (macroscopic_properties->m_mu_s == "constant") {
+            pml_mu_fp->setVal(macroscopic_properties->m_mu);
+        } else if (macroscopic_properties->m_mu_s == "parse_mu_function") {
+            macroscopic_properties->InitializeMacroMultiFabUsingParser(pml_mu_fp.get(),
+                getParser(macroscopic_properties->m_mu_parser), lev);
+        }
+
+    }
+
```

Comment:
General comment: the parser will read in based on (i,j,k) in the PML region.  Would it make sense to instead read the value from the parser ON the boundary instead of the location of the cell?  For AMR in the refined regions this would not be good.  I'm ok with leaving it as is, but then the material properites in the inputs file have to be carefully specified if a square of material is against the domain - we would have to extend this square out to infinity outside the domain, if that makes sense.

Reply:
YEs -- this is to be discussed.
The way it is right now -- it is easily extendable to pml inside the domain for MR.

But I agree that when we have a weird input with exponentially decreasing macroproperties, then for pml outside the domain, we need to take the boundary value.

For all our test-cases in the near future, this seems to be pretty linear or mostly even a jump condition with (zmid< and zmid>). Definitely addressing these corner cases is essential.
I am wondering if this can be done in a separate PR focusing only on corner cases for pml outside domain

Reply:
I think it's ok to leave as it by now.  It matches up with all our short and medium-term use cases.
For advanced cases down the line with more complex functional forms this can be revisited.